packages/vow/src/retrier.js

```diff
@@ -0,0 +1,137 @@
+import { M } from '@endo/patterns';
+import { PromiseWatcherI } from '@agoric/base-zone';
+import { isUpgradeDisconnection } from '@agoric/internal/src/upgrade-api.js';
+import { heapVowE, prepareVowTools } from '../vat.js';
+import { VowShape, toPassableCap } from './vow-utils.js';
+
+/**
+ * @import { Zone } from '@agoric/base-zone'
+ * @import {VowTools} from './tools.js'
+ */
+
+const RetrierI = M.interface('Retrier', {
+  retry: M.call().returns(),
+  getVow: M.call().returns(VowShape),
+  cancel: M.call(M.error()).returns(),
+});
+
+const RetrierShape = M.remotable('retrier');
+
+const RetrierAdminI = M.interface('RetrierAdmin', {
+  // modeled on getFlowForOutcomeVow
+  getRetrierForOutcomeVow: M.call(VowShape).returns(M.opt(RetrierShape)),
+});
+
+/**
+ * @param {Zone} zone
+ * @param {VowTools} [vowTools]
+ */
+export const prepareRetrierTools = (zone, vowTools = prepareVowTools(zone)) => {
+  const { makeVowKit, watch } = vowTools;
+  const retrierForOutcomeVowKey = zone.mapStore('retrierForOutcomeVow', {
+    keyShape: M.remotable('toPassableCap'),
+    valueShape: RetrierShape,
+  });
+
+  const makeRetrierKit = zone.exoClassKit(
+    'Retrier',
+    {
+      retrier: RetrierI,
+      watcher: PromiseWatcherI,
+    },
+    (target, optVerb, args) => {
+      const { vow, resolver } = makeVowKit();
+      return harden({
+        target,
+        optVerb,
+        args,
+        vow,
+        optResolver: resolver,
+      });
+    },
+    {
+      retrier: {
+        retry() {
+          const { state, facets } = this;
+          const { target, optVerb, args, optResolver } = state;
+          const { watcher } = facets;
+
+          if (optResolver === undefined) {
+            return;
+          }
+          // TODO `heapVowE` is likely too fragile under upgrade.
+          const p = optVerb
+            ? heapVowE(target)[optVerb](...args)
+            : heapVowE(target)(...args);
```

Comment:
This should not test for truthy vs. falsey `optVerb`:
```suggestion
          const p = optVerb === undefined
            ? heapVowE(target)(...args)
            : heapVowE(target)[optVerb](...args);
```

Reply:
Done